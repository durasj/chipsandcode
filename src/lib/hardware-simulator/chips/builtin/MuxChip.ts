import type { ChipPin } from '../Chip';
import type Chip from '../Chip';

/**
 * Mux chip implemented natively
 */
class MuxChip implements Chip {
  public readonly name = 'Mux' as string;

  private a: boolean[];
  private b: boolean[];
  private sel: boolean[];
  private out: boolean[];

  constructor(private width = 1) {
    this.a = new Array(width).fill(false);
    this.b = new Array(width).fill(false);
    this.sel = [false];
    this.out = new Array(width).fill(false);
  }

  setInput(name: 'a' | 'b' | 'sel', value: boolean[]) {
    this[name] = value;
  }

  getOutput() {
    return this.out;
  }

  getPins() {
    // Constructs map on the fly
    // Pin inspection is less performance sensitive than actual execution

    return new Map([
      ['a', { type: 'input', width: this.width, state: this.a, connections: [] }],
      ['b', { type: 'input', width: this.width, state: this.b, connections: [] }],
      ['sel', { type: 'input', width: 1, state: this.sel, connections: [] }],
      ['out', { type: 'output', width: this.width, state: this.out, connections: [] }],
    ]) satisfies Map<string, ChipPin>;
  }

  public run() {
    const pin = this.sel[0] ? this.b : this.a;
    this.out = pin;
  }
}

export default MuxChip;
