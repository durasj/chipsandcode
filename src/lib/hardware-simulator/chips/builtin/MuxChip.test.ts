import { describe, expect, it } from 'vitest';

import MuxChip from './MuxChip';

describe('Mux Chip', () => {
  it('Provides interface info', async () => {
    const chip = new MuxChip();

    expect(chip.name).toBe('Mux');
    expect(chip.getPins()).toMatchSnapshot();
  });

  it('Has correct logic', async () => {
    const chip = new MuxChip();

    const values = [
      [[false], [false], [false], [false]],
      [[true], [false], [false], [true]],
      [[false], [true], [false], [false]],
      [[true], [true], [false], [true]],
      [[false], [false], [true], [false]],
      [[true], [false], [true], [false]],
      [[false], [true], [true], [true]],
      [[true], [true], [true], [true]],
    ];

    values.forEach(([a, b, sel, out]) => {
      chip.setInput('a', a);
      chip.setInput('b', b);
      chip.setInput('sel', sel);
      chip.run();

      expect(chip.getOutput()).toStrictEqual(out);
    });
  });
});
