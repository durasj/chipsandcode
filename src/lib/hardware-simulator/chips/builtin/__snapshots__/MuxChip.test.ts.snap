// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Mux Chip > Provides interface info 1`] = `
Map {
  "a" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "b" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "sel" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "out" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "output",
    "width": 1,
  },
}
`;
